parsec/class/parsec_rbtree.h

```diff
@@ -0,0 +1,41 @@
+#ifndef PARSEC_RBTREE_H
+#define PARSEC_RBTREE_H
+
+#include "parsec/class/list_item.h"
+
+
+typedef enum parsec_rbtree_color_e { PARSEC_RBTREE_RED, PARSEC_RBTREE_BLACK } parsec_rbtree_color_e;
+
+typedef struct parsec_rbtree_node_t {
```

Comment:
Don't derive it from `parsec_list_item_t`, there are too many fields that are not only unnecessary, but will make a badly aligned object.

I would suggest 
```c
typedef struct parsec_rbtree_node_s {
    parsec_object_t  super;                         /**< Base object */
    volatile struct parsec_list_item_s* rbtree_left;  /**< Pointer to the left item */
    volatile struct parsec_list_item_s* rbtree_right;  /**< Pointer to the right item */
    parsec_rbtree_color_e color;
} parsec_rbtree_node_t;
```

Even with this the color enum will take 8 bytes.

Reply:
This layout was intentional. The current layout of `list_item_t` (without debugging fluff):

```C
typedef struct parsec_list_item_s {
    parsec_object_t  super;                         /**< A list item is a @ref parsec_internal_classes_object */
    volatile struct parsec_list_item_s* list_next;  /**< Pointer to the next item */
    volatile struct parsec_list_item_s* list_prev;  /**< Pointer to the previous item */
    int32_t aba_key;                                /**< This field is __very__ special and should be handled with extreme
                                                     *   care. It is used to avoid the ABA problem when atomic operations
                                                     *   are in use and we do not have support for 128 bits atomics.
                                                     */
} parsec_list_item_t;
```

The compiler will put the color right after the aba key, no gaps. The size is the same as what you suggest. By deriving from `parsec_list_item_t` these items can be put into a list or lifo without hacking around.

Reply:
The size of what I suggested will always be the same including in debug more, because all of the unused debugging information will not even be there. I don't see the rb item put in any lists as of now, but who knows maybe one day they will.

Reply:
We use the list item to store them in a lifo of free nodes to avoid reallocating them: https://github.com/ICLDisco/parsec/pull/710/files#diff-16341bb496e9d455f94a4e49f22ff87cd1894643e29364a6827b351785560a18R141